Grant/contracts/BuilderGrant.sol

```diff
@@ -0,0 +1,350 @@
+        season.grants[0].amount = grantsAmount[0];
+        season.grants[1].amount = grantsAmount[1];
+        season.grants[2].amount = grantsAmount[2];
+        season.claimStartedAt = uint40(block.timestamp);
+
+        // Reset ranker data if exists, and overwrites the old data
+        if (season.rankers.length > 0) {
+            delete season.rankers;
+        }
+        unchecked {
+            for (uint256 i = 0; i < fids.length; ++i) {
+                season.rankers.push(Ranker(fids[i], wallets[i], 0, [false, false, false]));
+            }
+        }
+
+        emit SetSeasonData(seasonId, fids.length, grantsAmount);
+    }
+
+    function getClaimableAmountByTop3(
+        uint256 seasonId,
+        uint256 ranking
+    ) public view validSeasonId(seasonId) returns (uint16) {
+        // Only the top 3 rankers (index: 0, 1, 2) can claim
+        if (ranking > 2) revert InvalidRankingParam();
+
+        // Check claim deadline
+        Season storage season = seasons[seasonId];
+        if (season.claimStartedAt + TOP3_CLAIM_DEADLINE < block.timestamp) revert ClaimDeadlineReached();
+
+        // Check if the grant has not been claimed yet
+        if (season.grants[ranking].claimedType != 0) revert AlreadyClaimed();
+
+        return season.grants[ranking].amount;
+    }
+
+    function claimByTop3(uint256 seasonId, uint256 ranking, uint8 claimType) external validSeasonId(seasonId) {
+        // Basic validations are done here
+        uint16 claimableAmount = getClaimableAmountByTop3(seasonId, ranking);
+
+        Season storage season = seasons[seasonId];
+        address msgSender = _msgSender();
+
+        // Check msg.sender is the winner
+        if (season.rankers[ranking].wallet != msgSender) revert PermissionDenied();
+
+        // claimType - 0: not claimed yet, 1: 100% self, 2: 50% donation, 3: 100% donation
+        uint16 amountForSelf;
+        uint16 amountForDonation;
+        if (claimType == 1) {
+            // 100% self
+            amountForSelf = claimableAmount;
+        } else if (claimType == 2) {
+            // 50% donation
+            amountForSelf = amountForDonation = claimableAmount / 2;
+        } else if (claimType == 3) {
+            // 100% donation
+            amountForDonation = claimableAmount;
+        } else {
+            revert InvalidClaimType();
+        }
+
+        // Set ranker claimed
+        season.grants[ranking].claimedType = claimType;
+
+        // Send Mini Buildings reward
+        if (amountForSelf > 0) {
+            season.rankers[ranking].claimedAmount += amountForSelf;
+            season.totalClaimed += amountForSelf;
+            if (!_mintBuildings(amountForSelf, msgSender)) revert MintBuildingsFailed();
+        }
+
+        // Set donations
+        if (amountForDonation > 0) {
+            unchecked {
+                for (uint256 i = 3; i < amountForDonation + 3; ++i) {
+                    season.rankers[i].donationReceived[ranking] = true;
+                }
```

Comment:
1. MEDIUM: When `season.rankers.length < amountForDonation + 3`, this will write beyond the length of the array.

For example, 10 rankers and a top grant of 20 donated 100%: the loop will iterate `i` from `[3..22]` even though `season.rankers` only has 10 entries.

2. GAS: Do you need to track `donationReceived` for every ranker? Can't you just track it for `ranking` e.g. `season.donationReceived[3]`?